test/puppetlabs/puppetdb/command_test.clj

```diff
@@ -1776,3 +1778,54 @@
             ;; fast enough, initially (at least), the rates can be zero.
             (is (= (+ 3 (:total start-global)) (:total global)))
             (is (= (+ 3 (or (:total start-cmd) 0)) (:total cmd)))))))))
+
+
+;; Test mitigation of
+;; https://bugs.openjdk.java.net/browse/JDK-8176254 i.e. handling of
+;; an unexpected in-flight garbage collection (via the scheduling
+;; bug) during stop.  For now, just test that if a gc is in flight,
+;; we wait on it if it doesn't take too long, and we proceed anyway
+;; if it does.
+
+(defn test-stop-with-delayed-commands [infinite-delay? expected-msg-rx]
+  (with-log-output log-output
+    (let [msg-count 3
+          enq-latch (java.util.concurrent.CountDownLatch. msg-count)
+          enq-proceed (promise)]
+      (with-redefs [cmd/schedule-msg-after (fn [ms f _] (f))  ;; no delay
+                    cmd/enqueue-delayed-message (fn [& args]
+                                                  (.countDown enq-latch)
+                                                  @enq-proceed
+                                                  ;; do nothing
+                                                  nil)]
+        (with-pdb-with-no-gc
+          (let [pdb (get-service *server* :PuppetDBServer)
+                q (-> pdb cli-svc/shared-globals :q)
+                dispatcher (get-service *server* :PuppetDBCommandDispatcher)
+                stop-status (-> dispatcher service-context :stop-status)
+                v5-catalog (get-in wire-catalogs [5 :empty])
+                cmdreqs (repeatedly msg-count #(catalog->command-req 5 v5-catalog))
+                dummy-ex (Exception.)]
+            (doseq [cmdreq cmdreqs
+                    :let [cmdref (queue/store-command q cmdreq)
+                          {:keys [command version]} cmdref]]
+              ;; schedule-delayed-message assumes the metrics already exist
+              (cmd/create-metrics-for-command! command version)
+              (utils/noisy-future
+               (cmd/schedule-delayed-message cmdref (Exception.)
```

Comment:
maybe the exception here was supposed to be the dummy-ex?

Reply:
Yeah, just removed it -- decided it didn't add enough information to be worth the indirection.